oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/api/binary/BinaryHttpUpload.java

```diff
@@ -0,0 +1,163 @@
+ * one which can be smaller.
+ *
+ * <p>
+ * Below is the detailed algorithm for the remote client.
+ *
+ * <p>
+ * The following variables are used:
+ * <ul>
+ *     <li><code>fileSize</code>: the actual binary size (must be known at this point)</li>
+ *     <li><code>minPartSize</code>: the value from {@link #getMinPartSize()}</li>
+ *     <li><code>maxPartSize</code>: the value from {@link #getMaxPartSize()}</li>
+ *     <li><code>partURLCount</code>: the number of entries in {@link #getURLParts()}</li>
+ *     <li><code>partURLs</code>: the entries in {@link #getURLParts()}</li>
+ *     <li><code>partSize</code>: the part size to be used in the upload (to be determined in the algorithm)</li>
+ * </ul>
+ *
+ * Steps:
+ * <ol>
+ *     <li>if (fileSize divided by maxPartSize) is larger than partURLCount, then the client cannot proceed
+ *     and will have to request a new set of URLs with the right fileSize as maxSize</li>
+ *     <li>if fileSize is smaller than minPartSize, then take the first provided part URL to upload
+ *     the entire binary, with partSize = fileSize</li>
+ *     <li>
+ *         (optional) if the client has more information to optimize, the partSize
+ *         can be chosen, under the condition that all of these are true for the partSize:
+ *         <ol>
+ *             <li>larger than minPartSize</li>
+ *             <li>smaller or equal than maxPartSize (unless it is -1 = unlimited)</li>
+ *             <li>larger than fileSize divided by partURLCount</li>
+ *         </ol>
+ *     </li>
+ *     <li>otherwise all part URLs are to be used and the partSize = fileSize divided by partURLCount
+ *     (integer division, discard modulo which will be the last part)</li>
+ *     <li>upload: segment the binary into partSize, for each segment take the next URL from partURLs
+ *     (strictly in order!), proceed with a standard HTTP PUT for each, and for the last part use whatever
+ *     segment size is left</li>
+ *     <li>if a segment fails during upload, retry (up to a certain time out)</li>
+ *     <li>after the upload has finished successfully, notify the application, for example through a
+ *     complete request, passing the {@link #getUploadToken() upload token}, and the application will have to
+ *     call {@link HttpBinaryProvider#completeHttpUpload(String)} with the token</li>
+ * </ol>
+ *
+ * <h2>JSON view</h2>
+ *
+ * A JSON representation of this interface as passed back to a remote client might look like this:
+ * <pre>
+ * {
+ *     "uploadToken": "aaaa-bbbb-cccc-dddd-eeee-ffff-gggg-hhhh",
+ *     "minPartSize": 10485760,
+ *     "maxPartSize": 104857600,
+ *     "partURLs": [
+ *         "http://server.com/upload/1",
+ *         "http://server.com/upload/2",
+ *         "http://server.com/upload/3",
+ *         "http://server.com/upload/4"
+ *     ]
+ * }
+ * </pre>
+ * </p>
+ *
+ */
+public interface BinaryHttpUpload {
```

Comment:
Should we rename this to `HttpBinaryUpload` to be consistent with `HttpBinaryProvider`?

I actually casually used this name in another comment before I realized it‘s the other way around.

Reply:
If we do that, we'd probably want to do that also with `BlobHttpUpload` and `DataRecordHttpUpload` for consistency.